src/Http/Middleware/ProxyAddress.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\Http\Middleware;
+
+use Flarum\Http\Exception\ProxyNotAllowedException;
+use Illuminate\Support\Arr;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface as Middleware;
+use Psr\Http\Server\RequestHandlerInterface;
+
+class ProxyAddress implements Middleware
+{
+    /**
+     * @var bool
+     */
+    protected $enabled;
+
+    /**
+     * @var array
+     */
+    protected $allowedAddresses;
+
+    /**
+     * @param bool $enabled
+     * @param array $allowedAddresses
+     */
+    public function __construct($enabled, $allowedAddresses)
+    {
+        $this->enabled = $enabled;
+        $this->allowedAddresses = $allowedAddresses;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $ipAddress = Arr::get($request->getServerParams(), 'REMOTE_ADDR', '127.0.0.1');
+
+        if ($this->enabled) {
+            if (in_array($ipAddress, $this->allowedAddresses)) {
+                // standard header for proxies, see: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For
+                $ipAddress = Arr::get($request->getServerParams(), 'X_FORWARDED_FOR', $ipAddress);
+                $ipAddress = Arr::get($request->getServerParams(), 'HTTP_CLIENT_IP', $ipAddress);
+            } else {
+                throw new ProxyNotAllowedException("The used proxy isn't allowed to connect!");
```

Comment:
We generally like to avoid hardcoding text in any particular language where possible. In this case though, is the exception message really neccessary? Injecting a translator instance seems... excessive when this exception is only raised in one instance.

Reply:
The only service/Proxy that I can think of that set's multiple headers at the same time is Cloudflare, but that's the X-Forwarded-For header and then a custom Cloudflare-IP header if I remember correctly. Officially there are only 3 proxy headers that are standard/matter for IPs and those are:

- Forwarded (not part of the HTTP standards, but is well known and used commonly)
- X-Forwarded-For
- Via

Reply:
> We generally like to avoid hardcoding text in any particular language where possible. In this case though, is the exception message really neccessary? Injecting a translator instance seems... excessive when this exception is only raised in one instance.

The exception message is not really neccessary, the exception name itself might be enough.

Reply:
Yeah let's go that route, thanks!

Reply:
@x7airworker I've been thinking some more about this, and I'm still a bit uneasy with hardcoding the headers so definitively (that stackoverflow link you shared also had some alternative options). Maybe it might be worth allowing a list of headers to be provided in the config.php, but defaulting to these 3? Of course, if the Forwarded RFC becomes standard, this will need to be revisited, but I don't think that's something we should worry about right now.

Aside from the exception message, I think that's my last comment on this, after that it should be good to merge. Thank you so much for implementing this, and I definitely hope to see more of your PRs in the future :grin:

Reply:
Right now I've implemented the biggest 3, if we've missed something, someone will sometime open an issue about some edge case header.
And the middleware can still be overwritten by extensions.

Reply:
Alright, that's fair! Not something worth bikeshedding about :D